File: messaging_llm_bot/whatsapp_bridge.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const path = require("path");
const process = require("process");
const qrcode = require("qrcode-terminal");
const { Client, LocalAuth, MessageMedia } = require("whatsapp-web.js");

function printUsage() {
  console.error(
    "Usage: node messaging_llm_bot/whatsapp_bridge.js <pair|receive|send> [options]\n" +
      "Commands:\n" +
      "  pair --session <dir>\n" +
      "  receive --session <dir> [--state-file <path>] [--media-dir <path>]\n" +
      "  send --session <dir> --recipient <id> [--message <text>] [--attachment <path>]"
  );
}

function parseArgs(argv) {
  if (argv.length === 0) {
    printUsage();
    process.exit(2);
  }
  const command = argv[0];
  const options = {
    session: "data/whatsapp-default",
    stateFile: null,
    mediaDir: null,
    recipient: "",
    message: "",
    attachment: "",
    headless: true,
    browserPath: "",
    readyTimeoutSeconds: 45,
  };

  for (let index = 1; index < argv.length; index += 1) {
    const key = argv[index];
    const value = argv[index + 1];
    if (key === "--headful") {
      options.headless = false;
      continue;
    }
    if (key === "--headless") {
      options.headless = value !== "false";
      index += 1;
      continue;
    }
    if (!key.startsWith("--")) {
      throw new Error(`Unexpected argument: ${key}`);
    }
    if (value === undefined || value.startsWith("--")) {
      throw new Error(`Missing value for ${key}`);
    }
    switch (key) {
      case "--session":
        options.session = value;
        break;
      case "--state-file":
        options.stateFile = value;
        break;
      case "--media-dir":
        options.mediaDir = value;
        break;
      case "--recipient":
        options.recipient = value;
        break;
      case "--message":
        options.message = value;
        break;
      case "--attachment":
        options.attachment = value;
        break;
      case "--browser-path":
        options.browserPath = value;
        break;
      case "--ready-timeout-seconds":
        options.readyTimeoutSeconds = Number(value);
        break;
      default:
        throw new Error(`Unknown option: ${key}`);
    }
    index += 1;
  }

  if (!Number.isFinite(options.readyTimeoutSeconds) || options.readyTimeoutSeconds <= 0) {
    throw new Error("--ready-timeout-seconds must be > 0");
  }

  return { command, options };
}

function ensureDir(targetPath) {
  fs.mkdirSync(targetPath, { recursive: true });
}

function resolvePath(targetPath) {
  return path.resolve(process.cwd(), targetPath);
}

function sanitizeName(value) {
  return value.replace(/[^a-zA-Z0-9_-]+/g, "-").replace(/^-+|-+$/g, "") || "default";
}

function sessionInfo(sessionPath) {
  const absoluteSessionPath = resolvePath(sessionPath);
  const parentDir = path.dirname(absoluteSessionPath);
  const clientId = sanitizeName(path.basename(absoluteSessionPath));
  ensureDir(parentDir);
  ensureDir(absoluteSessionPath);
  return {
    absoluteSessionPath,
    parentDir,
    clientId,
  };
}

function defaultStateFile(sessionPath) {
  return `${resolvePath(sessionPath)}.receive-state.json`;
}

function defaultMediaDir(sessionPath) {
  return `${resolvePath(sessionPath)}-media`;
}

function loadState(stateFile) {
  try {
    return JSON.parse(fs.readFileSync(stateFile, "utf8"));
  } catch (error) {
    return { seen_message_ids: [] };
  }
}

function saveState(stateFile, state) {
  ensureDir(path.dirname(stateFile));
  fs.writeFileSync(stateFile, JSON.stringify(state, null, 2));
}

function withTimeout(promise, ms, label) {
  let timer = null;
  const timeout = new Promise((_, reject) => {
    timer = setTimeout(() => reject(new Error(`${label} timed out after ${ms}ms`)), ms);
  });
  return Promise.race([promise, timeout]).finally(() => {
    if (timer) {
      clearTimeout(timer);
    }
  });
}

function createClient(options) {
  const { parentDir, clientId } = sessionInfo(options.session);
  const puppeteer = {
    headless: options.headless,
    args: ["--no-sandbox", "--disable-setuid-sandbox", "--disable-dev-shm-usage"],
  };
  if (options.browserPath) {
    puppeteer.executablePath = resolvePath(options.browserPath);
  }
  return new Client({
    authStrategy: new LocalAuth({
      clientId,
      dataPath: parentDir,
    }),
    puppeteer,
  });
}

function waitForReady(client, timeoutMs) {
  return withTimeout(
    new Promise((resolve, reject) => {
      let qrPrinted = false;
      client.on("qr", (qrValue) => {
        qrPrinted = true;
        qrcode.generate(qrValue, { small: true });
        console.error("Scan the QR in WhatsApp -> Settings -> Linked Devices -> Link a Device");
      });
      client.on("authenticated", () => {
        console.error("WhatsApp authenticated.");
      });
      client.on("ready", () => {
        if (!qrPrinted) {
          console.error("WhatsApp session already linked.");
        }
        resolve();
      });
      client.on("auth_failure", (message) => {
        reject(new Error(`WhatsApp authentication failed: ${message || "unknown error"}`));
      });
      client.on("disconnected", (reason) => {
        reject(new Error(`WhatsApp disconnected before ready: ${reason || "unknown reason"}`));
      });
      client.initialize().catch(reject);
    }),
    timeoutMs,
    "WhatsApp session initialization"
  );
}

function normalizePhoneId(raw) {
  const digits = String(raw || "")
    .trim()
    .replace(/[^+\d]/g, "");
  if (!digits) {
    return "";
  }
  return `${digits.replace(/^\+/, "")}@c.us`;
}

function conversationIdForChat(chat) {
  if (chat.isGroup) {
    const name = String(chat.name || "group").replace(/\|/g, "/").trim() || "group";
    return `group:${name}|${chat.id._serialized}`;
  }
  return chat.id._serialized;
}

function senderIdForMessage(message, chat) {
  if (chat.isGroup) {
    return message.author || message.from || "";
  }
  return message.from || chat.id._serialized;
}

function sentAtForMessage(message) {
  if (!message.timestamp) {
    return null;
  }
  return new Date(message.timestamp * 1000).toISOString();
}

function mediaFilename(message, extension) {
  const baseName = message.id && message.id.id ? message.id.id : `media-${Date.now()}`;
  return `${baseName}.${extension || "bin"}`;
}

async function downloadAttachment(message, mediaDir) {
  if (!message.hasMedia) {
    return null;
  }
  const media = await message.downloadMedia();
  if (!media || !media.data) {
    return null;
  }
  ensureDir(mediaDir);
  const extension = media.mimetype && media.mimetype.includes("/")
    ? media.mimetype.split("/")[1].split(";")[0]
    : "bin";
  const filename = media.filename || mediaFilename(message, extension);
  const filePath = path.join(mediaDir, filename);
  fs.writeFileSync(filePath, Buffer.from(media.data, "base64"));
  return {
    path: filePath,
    name: filename,
    mime_type: media.mimetype || "application/octet-stream",
  };
}

async function commandPair(options) {
  const client = createClient(options);
  try {
    await waitForReady(client, options.readyTimeoutSeconds * 1000);
    console.error(`Linked session stored under ${resolvePath(options.session)}`);
  } finally {
    await client.destroy().catch(() => {});
  }
}

async function commandReceive(options) {
  const client = createClient(options);
  const stateFile = resolvePath(options.stateFile || defaultStateFile(options.session));
  const mediaDir = resolvePath(options.mediaDir || defaultMediaDir(options.session));
  const state = loadState(stateFile);
  const seen = new Set(Array.isArray(state.seen_message_ids) ? state.seen_message_ids : []);

  try {
    await waitForReady(client, options.readyTimeoutSeconds * 1000);
    const chats = await client.getChats();
    const messages = [];

    for (const chat of chats) {
      const unreadCount = Number(chat.unreadCount || 0);
      if (unreadCount <= 0) {
        continue;
      }
      const recent = await chat.fetchMessages({ limit: Math.max(unreadCount, 20) });
      const unreadMessages = recent.filter((message) => !message.fromMe);
      for (const message of unreadMessages) {
        const messageId = message.id && message.id._serialized ? message.id._serialized : "";
        if (!messageId || seen.has(messageId)) {
          continue;
        }
        const attachment = await downloadAttachment(message, mediaDir);
        const text = typeof message.body === "string" ? message.body.trim() : "";
        if (!text && !attachment) {
          seen.add(messageId);
          continue;
        }
        const contact = await message.getContact();
        messages.push({
          conversation_id: conversationIdForChat(chat),
          sender_id: senderIdForMessage(message, chat),
          sender_name: contact.pushname || contact.name || contact.shortName || null,
          sender_contact: contact.number || contact.pushname || contact.name || senderIdForMessage(message, chat),
          text: text || null,
          sent_at: sentAtForMessage(message),
          attachments: attachment ? [attachment] : [],
        });
        seen.add(messageId);
      }
    }

    const seenList = Array.from(seen);
    state.seen_message_ids = seenList.slice(-2000);
    saveState(stateFile, state);
    process.stdout.write(JSON.stringify({ messages }));
  } finally {
    await client.destroy().catch(() => {});
  }
}

async function resolveRecipient(client, recipient) {
  const trimmed = String(recipient || "").trim();
  if (!trimmed) {
    throw new Error("--recipient is required");
  }
  if (trimmed.startsWith("group:") && trimmed.includes("|")) {
    return trimmed.split("|").slice(-1)[0];
  }
  if (trimmed.includes("@")) {
    return trimmed;
  }
  const directId = normalizePhoneId(trimmed);
  if (!directId) {
    throw new Error(`Unsupported recipient value: ${trimmed}`);
  }
  const numberId = await client.getNumberId(trimmed.replace(/[^\d]/g, ""));
  if (numberId && numberId._serialized) {
    return numberId._serialized;
  }
  return directId;
}

async function commandSend(options) {
  if (!options.recipient) {
    throw new Error("--recipient is required");
  }
  if (!options.message && !options.attachment) {
    throw new Error("either --message or --attachment is required");
  }
  const client = createClient(options);
  try {
    await waitForReady(client, options.readyTimeoutSeconds * 1000);
    const chatId = await resolveRecipient(client, options.recipient);
    if (options.attachment) {
      const attachmentPath = resolvePath(options.attachment);
      if (!fs.existsSync(attachmentPath)) {
        throw new Error(`Attachment file not found: ${attachmentPath}`);
      }
      const media = MessageMedia.fromFilePath(attachmentPath);
      await client.sendMessage(chatId, media, { caption: options.message || "" });
      return;
    }
    await client.sendMessage(chatId, options.message);
  } finally {
    await client.destroy().catch(() => {});
  }
}

async function main() {
  try {
    const { command, options } = parseArgs(process.argv.slice(2));
    if (command === "pair") {
      await commandPair(options);
      return;
    }
    if (command === "receive") {
      await commandReceive(options);
      return;
    }
    if (command === "send") {
      await commandSend(options);
      return;
    }
    throw new Error(`Unknown command: ${command}`);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error(message);
    process.exit(1);
  }
}

main();

File: messaging_llm_bot/whatsapp_bridge.py
from __future__ import annotations

import argparse
import json
import sys
import time
from pathlib import Path
from typing import Any
from urllib.parse import quote

WHATSAPP_WEB_URL = "https://web.whatsapp.com/"


class WhatsAppBridgeError(RuntimeError):
    """Raised when WhatsApp Web automation cannot complete."""


def _ensure_dir(path: Path) -> None:
    path.mkdir(parents=True, exist_ok=True)


def _resolve_path(value: str) -> Path:
    return Path(value).expanduser().resolve()


def _default_state_file(session_path: str) -> str:
    return f"{_resolve_path(session_path)}.receive-state.json"


def _default_media_dir(session_path: str) -> str:
    return f"{_resolve_path(session_path)}-media"


def _load_state(path: Path) -> dict[str, Any]:
    try:
        raw = json.loads(path.read_text(encoding="utf-8"))
    except Exception:
        return {"seen_message_ids": []}
    if not isinstance(raw, dict):
        return {"seen_message_ids": []}
    seen = raw.get("seen_message_ids")
    if not isinstance(seen, list):
        seen = []
    return {"seen_message_ids": [str(item) for item in seen]}


def _save_state(path: Path, state: dict[str, Any]) -> None:
    _ensure_dir(path.parent)
    path.write_text(json.dumps(state, indent=2, sort_keys=True), encoding="utf-8")


def _runtime_imports():
    try:
        from playwright.sync_api import Error as PlaywrightError  # type: ignore
        from playwright.sync_api import TimeoutError as PlaywrightTimeoutError  # type: ignore
        from playwright.sync_api import sync_playwright  # type: ignore
    except Exception as exc:  # pragma: no cover - depends on local runtime
        raise WhatsAppBridgeError(
            "Playwright is required for the pure-Python WhatsApp bridge. "
            "Install it in this Python environment and run 'python3 -m playwright install chromium'."
        ) from exc
    return sync_playwright, PlaywrightError, PlaywrightTimeoutError


def _launch_context(
    *,
    session_dir: Path,
    headless: bool,
    browser_path: str,
    timeout_ms: int,
):
    sync_playwright, PlaywrightError, _ = _runtime_imports()
    _ensure_dir(session_dir)
    playwright = sync_playwright().start()
    launch_kwargs: dict[str, Any] = {
        "headless": headless,
        "args": ["--disable-blink-features=AutomationControlled"],
    }
    if browser_path:
        launch_kwargs["executable_path"] = str(_resolve_path(browser_path))
    try:
        context = playwright.chromium.launch_persistent_context(str(session_dir), **launch_kwargs)
    except PlaywrightError as exc:  # pragma: no cover - depends on local runtime
        playwright.stop()
        raise WhatsAppBridgeError(f"Failed to launch Chromium for WhatsApp: {exc}") from exc
    context.set_default_timeout(timeout_ms)
    page = context.pages[0] if context.pages else context.new_page()
    return playwright, context, page


def _open_whatsapp(page: Any, timeout_ms: int) -> None:
    page.goto(WHATSAPP_WEB_URL, wait_until="domcontentloaded", timeout=timeout_ms)


def _wait_for_login(page: Any, timeout_ms: int, *, interactive_wait_ms: int = 0) -> None:
    _, _, PlaywrightTimeoutError = _runtime_imports()
    ready_selectors = [
        "div[aria-label='Chat list']",
        "div[data-testid='chat-list']",
        "div[role='grid']",
        "div[data-testid='pane-side']",
    ]
    ready_selector = ", ".join(ready_selectors)
    try:
        page.wait_for_selector(ready_selector, timeout=timeout_ms)
        return
    except PlaywrightTimeoutError:
        if interactive_wait_ms > 0:
            print(
                "WhatsApp Web is not linked yet. Scan the QR in the open browser window:",
                file=sys.stderr,
            )
            print("WhatsApp -> Settings -> Linked Devices -> Link a Device", file=sys.stderr)
            page.wait_for_selector(ready_selector, timeout=interactive_wait_ms)
            return
        raise WhatsAppBridgeError(
            "WhatsApp Web is not ready. Run the pair command in headful mode and complete Linked Devices login."
        )


def _conversation_lookup_label(recipient: str) -> str:
    if recipient.startswith("group:"):
        _, _, tail = recipient.partition(":")
        name, _, raw_id = tail.partition("|")
        return name.strip() or raw_id.strip()
    if recipient.endswith("@c.us") or recipient.endswith("@g.us"):
        return recipient.split("@", 1)[0]
    return recipient.strip()


def _phone_digits(value: str) -> str:
    return "".join(ch for ch in value if ch.isdigit())


def _install_store(page: Any) -> None:
    page.evaluate(
        """
        () => {
          if (window.__BUDUCCA_WA_STORE) {
            return true;
          }
          const chunkName = Object.keys(window).find((key) => key.startsWith("webpackChunk"));
          if (!chunkName || !Array.isArray(window[chunkName])) {
            throw new Error("WhatsApp webpack runtime not found");
          }
          const chunk = window[chunkName];
          let webpackRequire = null;
          chunk.push([[`buducca_${Date.now()}`], {}, (req) => { webpackRequire = req; }]);
          if (!webpackRequire) {
            throw new Error("WhatsApp webpack require is unavailable");
          }
          const modules = [];
          for (const moduleId of Object.keys(webpackRequire.m || {})) {
            try {
              modules.push(webpackRequire(moduleId));
            } catch (error) {
            }
          }
          const find = (predicate) => modules.find((mod) => {
            try {
              return Boolean(predicate(mod));
            } catch (error) {
              return false;
            }
          });
          const storeContainer = find((mod) => mod?.Chat?.models && mod?.Msg?.models);
          const widFactory = find((mod) => typeof mod?.createWid === "function");
          const sendTextModule = find((mod) => typeof mod?.sendTextMsgToChat === "function");
          window.__BUDUCCA_WA_STORE = {
            Chat: storeContainer?.Chat || null,
            Msg: storeContainer?.Msg || null,
            WidFactory: widFactory || null,
            sendTextMsgToChat: sendTextModule?.sendTextMsgToChat || null,
          };
          if (!window.__BUDUCCA_WA_STORE.Chat || !window.__BUDUCCA_WA_STORE.Msg) {
            throw new Error("WhatsApp data store is unavailable");
          }
          return true;
        }
        """
    )


def _get_unseen_messages(page: Any, seen_message_ids: list[str]) -> list[dict[str, Any]]:
    _install_store(page)
    result = page.evaluate(
        """
        (seenIds) => {
          const seen = new Set(seenIds || []);
          const store = window.__BUDUCCA_WA_STORE;
          const chats = Array.isArray(store.Chat?.models) ? store.Chat.models : [];
          const messages = [];
          for (const chat of chats) {
            const chatMessages = Array.isArray(chat?.msgs?.models) ? chat.msgs.models : [];
            for (const msg of chatMessages) {
              const id = msg?.id?._serialized || null;
              if (!id || seen.has(id) || msg?.fromMe || msg?.isNotification) {
                continue;
              }
              const chatId = chat?.id?._serialized || "";
              const chatName = String(chat?.formattedTitle || chat?.name || "chat").replaceAll("|", "/").trim() || "chat";
              const isGroup = Boolean(chat?.isGroup);
              const text = String(msg?.body || msg?.caption || "").trim();
              const type = String(msg?.type || "").trim();
              const normalizedText = text || (type && type !== "chat" ? `[Attachment: ${type}]` : "");
              if (!normalizedText) {
                continue;
              }
              messages.push({
                message_id: id,
                conversation_id: isGroup ? `group:${chatName}|${chatId}` : chatId,
                sender_id: msg?.author || msg?.from || chatId,
                sender_name: msg?._data?.notifyName || "",
                sender_contact: msg?._data?.notifyName || msg?.author || msg?.from || chatName,
                text: normalizedText,
                sent_at: msg?.t ? new Date(msg.t * 1000).toISOString() : null,
              });
            }
          }
          messages.sort((left, right) => {
            const leftTime = left.sent_at || "";
            const rightTime = right.sent_at || "";
            return leftTime.localeCompare(rightTime);
          });
          return messages;
        }
        """,
        seen_message_ids,
    )
    if not isinstance(result, list):
        return []
    return [item for item in result if isinstance(item, dict)]


def _open_chat_via_search(page: Any, recipient: str, timeout_ms: int) -> None:
    label = _conversation_lookup_label(recipient)
    phone = _phone_digits(label)
    if phone and recipient == label:
        page.goto(f"{WHATSAPP_WEB_URL}send?phone={quote(phone)}", wait_until="domcontentloaded", timeout=timeout_ms)
    search_box = page.locator("div[contenteditable='true'][role='textbox']").first
    search_box.wait_for(timeout=timeout_ms)
    search_box.click()
    search_box.fill("")
    search_box.type(label, delay=20)
    time.sleep(0.5)
    result = page.locator("[role='listitem'], [data-testid='cell-frame-container']").first
    result.wait_for(timeout=timeout_ms)
    result.click()


def _send_text(page: Any, message: str, timeout_ms: int) -> None:
    input_box = page.locator("footer div[contenteditable='true'][role='textbox']").last
    input_box.wait_for(timeout=timeout_ms)
    input_box.click()
    input_box.type(message, delay=10)
    input_box.press("Enter")


def _send_attachment(page: Any, attachment: Path, caption: str, timeout_ms: int) -> None:
    attach_input = page.locator("input[type='file']").first
    if attach_input.count() == 0:
        attach_button = page.locator("[data-testid='clip'], span[data-icon='plus-rounded']").first
        attach_button.wait_for(timeout=timeout_ms)
        attach_button.click()
        attach_input = page.locator("input[type='file']").first
    attach_input.set_input_files(str(attachment))
    caption_box = page.locator("div[contenteditable='true'][role='textbox']").last
    caption_box.wait_for(timeout=timeout_ms)
    if caption:
        caption_box.click()
        caption_box.type(caption, delay=10)
    send_button = page.locator("[data-testid='send'], span[data-icon='send']").last
    send_button.wait_for(timeout=timeout_ms)
    send_button.click()


def command_pair(args: argparse.Namespace) -> int:
    headless = args.headless
    if args.command == "pair" and not args.headless_explicit:
        headless = False
    playwright, context, page = _launch_context(
        session_dir=_resolve_path(args.session),
        headless=headless,
        browser_path=args.browser_path,
        timeout_ms=args.ready_timeout_seconds * 1000,
    )
    try:
        _open_whatsapp(page, args.ready_timeout_seconds * 1000)
        _wait_for_login(
            page,
            args.ready_timeout_seconds * 1000,
            interactive_wait_ms=args.signup_wait_seconds * 1000,
        )
        print("WhatsApp session is linked.", file=sys.stderr)
        return 0
    finally:
        context.close()
        playwright.stop()


def command_receive(args: argparse.Namespace) -> int:
    playwright, context, page = _launch_context(
        session_dir=_resolve_path(args.session),
        headless=args.headless,
        browser_path=args.browser_path,
        timeout_ms=args.ready_timeout_seconds * 1000,
    )
    try:
        _open_whatsapp(page, args.ready_timeout_seconds * 1000)
        _wait_for_login(page, args.ready_timeout_seconds * 1000)
        state_path = _resolve_path(args.state_file or _default_state_file(args.session))
        state = _load_state(state_path)
        raw_messages = _get_unseen_messages(page, state.get("seen_message_ids", []))
        seen_ids = set(state.get("seen_message_ids", []))
        messages: list[dict[str, Any]] = []
        for item in raw_messages:
            message_id = str(item.get("message_id") or "").strip()
            if message_id:
                seen_ids.add(message_id)
            payload = {key: value for key, value in item.items() if key != "message_id"}
            messages.append(payload)
        _save_state(state_path, {"seen_message_ids": sorted(seen_ids)})
        print(json.dumps({"messages": messages}))
        return 0
    finally:
        context.close()
        playwright.stop()


def command_send(args: argparse.Namespace) -> int:
    playwright, context, page = _launch_context(
        session_dir=_resolve_path(args.session),
        headless=args.headless,
        browser_path=args.browser_path,
        timeout_ms=args.ready_timeout_seconds * 1000,
    )
    try:
        _open_whatsapp(page, args.ready_timeout_seconds * 1000)
        _wait_for_login(page, args.ready_timeout_seconds * 1000)
        _open_chat_via_search(page, args.recipient, args.ready_timeout_seconds * 1000)
        attachment = Path(args.attachment).expanduser() if args.attachment else None
        if attachment:
            if not attachment.exists():
                raise WhatsAppBridgeError(f"Attachment file does not exist: {attachment}")
            _send_attachment(page, attachment.resolve(), args.message or "", args.ready_timeout_seconds * 1000)
        elif args.message:
            _send_text(page, args.message, args.ready_timeout_seconds * 1000)
        else:
            raise WhatsAppBridgeError("Either --message or --attachment must be provided")
        return 0
    finally:
        context.close()
        playwright.stop()


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        prog="whatsapp-bridge",
        description="Pure-Python WhatsApp Web bridge for BUDUCCA",
    )
    subparsers = parser.add_subparsers(dest="command", required=True)

    def add_common(subparser: argparse.ArgumentParser) -> None:
        subparser.add_argument("--session", default="data/whatsapp-default")
        subparser.add_argument("--browser-path", default="")
        subparser.add_argument("--ready-timeout-seconds", type=int, default=45)
        subparser.add_argument("--signup-wait-seconds", type=int, default=300)
        subparser.add_argument("--headless", dest="headless", action="store_true")
        subparser.add_argument("--headful", dest="headless", action="store_false")
        subparser.set_defaults(headless=True, headless_explicit=False)

    pair = subparsers.add_parser("pair", help="Open WhatsApp Web and wait for device linking")
    add_common(pair)
    pair.set_defaults(func=command_pair)

    receive = subparsers.add_parser("receive", help="Emit unread WhatsApp messages as JSON")
    add_common(receive)
    receive.add_argument("--state-file", default="")
    receive.add_argument("--media-dir", default=_default_media_dir("data/whatsapp-default"))
    receive.set_defaults(func=command_receive)

    send = subparsers.add_parser("send", help="Send a WhatsApp message or attachment")
    add_common(send)
    send.add_argument("--recipient", required=True)
    send.add_argument("--message", default="")
    send.add_argument("--attachment", default="")
    send.set_defaults(func=command_send)

    return parser


def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
    parser = build_parser()
    args = parser.parse_args(argv)
    raw_argv = argv if argv is not None else sys.argv[1:]
    args.headless_explicit = "--headless" in raw_argv or "--headful" in raw_argv
    return args


def main(argv: list[str] | None = None) -> int:
    args = parse_args(argv)
    try:
        return args.func(args)
    except WhatsAppBridgeError as exc:
        print(str(exc), file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
